refactor(core): Consolidates shouldTrigger* methods down to one

thePunderWoman opened the pull request:
This removes potential confusion by relying on one shouldTriggerDeferBlock method everywhere.

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Refactoring (no functional changes, no api changes)



## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

thePunderWoman committed:
refactor(core): Consolidates shouldTrigger* methods down to one

This cleans up the triggering code base and consolidates it down to one
function that outlines the logic. This also resolves the `hydrate when`
behavior issue.

fixes: #58709

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -183,14 +181,16 @@ export function ɵɵdefer(
  */
 export function ɵɵdeferWhen(rawValue: unknown) {
   const lView = getLView();
+  const tNode = getSelectedTNode();
+
+  if (!shouldTriggerDeferBlock(lView, tNode)) return;
```
I'd like to propose a slightly different API. We can have a function that would accept the necessary details (e.g. `lView`, `tNode`, etc) and also a trigger type and return whether those triggers should be attached, for example:

```
export function shouldAttachTrigger(triggerType: TriggerType, lView: LView, tNode: TNode): boolean {
  // ... determine whether a given trigger type should be activated ...
}
```

The actual usage would look like this:
```
export function ɵɵdeferWhen(rawValue: unknown) {
  const lView = getLView();
  const tNode = getSelectedTNode();

  if (shouldAttachTrigger(TriggerType.Regular, lView, tNode)) {
    // ... the code that attaches the trigger ...
  }
```

For a "hydrate" trigger:
```
export function ɵɵdeferHydrateWhen(rawValue: unknown) {
  const lView = getLView();
  const tNode = getSelectedTNode();

  if (shouldAttachTrigger(TriggerType.Hydrate, lView, tNode)) {
    // ... the code that attaches the trigger ...
  }
```

We'd need to ensure that the `TriggerType.Regular` and `TriggerType.Hydrate` are mutually exclusive, i.e. we can compute a value based on the rules mentioned in the code below and return true/false depending on a trigger type.

AndrewKushnir commented on packages/core/src/defer/instructions.ts:
```diff
@@ -342,19 +341,14 @@ export function ɵɵdeferHydrateOnIdle() {
 export function ɵɵdeferOnImmediate() {
   const lView = getLView();
   const tNode = getCurrentTNode()!;
-  const tView = lView[TVIEW];
```
It looks like this function is missing an early exit (similar to other functions).

thePunderWoman replied:
I did run into a problem with early exit functions. When we add those, we break our testing defer block config behaviors for some functions. We should probably revisit that logic.

AndrewKushnir commented on packages/core/src/defer/triggering.ts:
```diff
@@ -74,16 +74,17 @@ export function scheduleDelayedTrigger(
   const injector = lView[INJECTOR]!;
   const lDetails = getLDeferBlockDetails(lView, tNode);
-  const tDetails = getTDeferBlockDetails(lView[TVIEW], tNode);
+  const isServerRendered =
+    lDetails[SSR_BLOCK_STATE] !== null && lDetails[SSR_BLOCK_STATE] === DeferBlockState.Complete;
```
I guess the intention was to check for the `SSR_UNIQUE_ID` field value presence?
```suggestion
  const isServerRendered =
    lDetails[SSR_UNIQUE_ID] !== null && lDetails[SSR_BLOCK_STATE] === DeferBlockState.Complete;
```

AndrewKushnir replied:
I also think that there should be no conditions in this function and we should move the check to an instruction level (i.e. to a call site).

AndrewKushnir commented on packages/core/src/defer/triggering.ts:
```diff
@@ -471,69 +467,78 @@ function onDeferBlockCompletion(lDetails: LDeferBlockDetails, callback: VoidFunc
+  const lDetails = getLDeferBlockDetails(lView, tNode);
+  const tDetails = getTDeferBlockDetails(lView[TVIEW], tNode);
+  const incrementalHydrationEnabled = isIncrementalHydrationEnabled(lView[INJECTOR]!);
+  const hasHydrateTriggers =
+    lDetails[SSR_UNIQUE_ID] !== null ||
+    (tDetails.hydrateTriggers !== null && tDetails.hydrateTriggers.size > 0);
```
If we rely on the `tDetails.hydrateTriggers` presence, we'll make an implicit ordering dependency between `deferHydrate*` and `defer*` (non-hydrate) instructions, i.e. hydrate ones must be invoked first. This might be a problem for `deferHydrateWhen` instruction, which would end up in the "update" block of a template function, when the other functions (e.g. `deferOnIdle`) would be in the "create" block. I'm wondering if actually checking the `SSR_UNIQUE_ID` field value presence in `LDetails` should be sufficient. That would mean that this block was triggered on the server, thus it has "hydrate" triggers. 
```suggestion
  const hasHydrateTriggers = lDetails[SSR_UNIQUE_ID] !== null;
```

thePunderWoman replied:
You are correct, though we do implicitly ingest hydrate triggers first in the compiler for this very reason. Perhaps we should be consistent about processing hydrate triggers first everywhere, since we do actually need to know if they are present.

thePunderWoman committed:
refactor(compiler): Adds ingest and flags for defer details

This adds TDeferDetailsFlags to indicate the presence of hydration triggers, and any future flags we add to defer.

AndrewKushnir commented on packages/core/src/defer/triggering.ts:
```diff
@@ -74,16 +75,16 @@ export function scheduleDelayedTrigger(
-    const cleanupFn = scheduleFn(() => triggerDeferBlock(lView, tNode), injector);
-    storeTriggerCleanupFn(TriggerType.Regular, lDetails, cleanupFn);
-  }
+  // Only trigger the scheduled trigger on the browser
+  // since we don't want to delay the server response.
```
nit: this comment is obsolete, we can remove it.

AndrewKushnir commented on packages/core/src/defer/triggering.ts:
```diff
@@ -472,66 +483,50 @@ function onDeferBlockCompletion(lDetails: LDeferBlockDetails, callback: VoidFunc
+  // On the server:
   if (typeof ngServerMode !== 'undefined' && ngServerMode) {
-    return false;
+    // Only when incremental hydration is disabled OR no hydrate trigger is present for ɵɵdefer* (non-hydrate)
+    // OR always when incremental hydration is enabled for ɵɵdeferHydrate*
```
```suggestion
    // Regular triggers are activated on the server when:
    //  - Either Incremental Hydration is *not* enabled
    //  - Or Incremental Hydration is enabled, but a given block doesn't have "hydrate" triggers
```

AndrewKushnir commented on packages/platform-server/test/incremental_hydration_spec.ts:
```diff
@@ -1235,9 +1221,94 @@ describe('platform-server partial hydration integration', () => {
+
+      const article = appHostNode.querySelector('article');
+
+      expect(article.__ngDebugHydrationInfo__).toBeUndefined();
```
It'd be great to avoid reading `__ngDebugHydrationInfo__` directly, we can use the `readHydrationInfo` helper method for this.

We can also create a utility function in tests to verify whether a given node has been hydrated, for example:

```suggestion
      verifyElementWasHydrated(article);
```

The helper method itself may look like this:

```
function verifyElementWasHydrated(element: HTMLElement) {
  if (readHydrationInfo(element)?.status !== HydrationStatus.Hydrated) {
    fail('Hydration error: the node is *not* hydrated: ' + el.outerHTML);
  }
}
```

thePunderWoman replied:
There's another spot just a few lines above where I was also reading this. So I've created two helper functions, one that verifies hydration and the other that verifies the inverse.

AndrewKushnir commented on packages/platform-server/test/incremental_hydration_spec.ts:
```diff
@@ -1235,9 +1221,94 @@ describe('platform-server partial hydration integration', () => {
+
+      expect(registry.has('d0')).toBeFalsy();
+      expect(appHostNode.outerHTML).toContain('<span id="test">start</span>');
+    }, 100000);
```
```suggestion
    });
```

AndrewKushnir commented on packages/platform-server/test/incremental_hydration_spec.ts:
```diff
@@ -1235,9 +1221,94 @@ describe('platform-server partial hydration integration', () => {
+      const clickEvent = new CustomEvent('click');
+      testElement.dispatchEvent(clickEvent);
+
+      await timeout(1000); // wait for defer blocks to resolve
```
Do we need to wait for 1 second here? We can probably replace this with:
```suggestion
      await allPendingDynamicImports();
```

AndrewKushnir reviewed (approved):
LGTM, just a few minor comments (in docs and tests).

pkozlowski-opensource commented:
This PR was merged into the repository by commit 30891d8dec1060c4471563346deb60be4946c807.

The changes were merged into the following branches: main, 19.0.x
